doc/rasctl.1
.TH rascsi 1
.SH NAME
rasctl \- Sends management commands to the rascsi process
.SH SYNOPSIS
.B rasctl
\fB\-e\fR |
\fB\-l\fR |
\fB\-m\fR |
\fB\-o\fR |
\fB\-s\fR |
\fB\-v\fR |
\fB\-D\fR |
\fB\-I\fR |
\fB\-L\fR |
\fB\-O\fR |
\fB\-P\fR |
\fB\-T\fR |
\fB\-V\fR |
\fB\-X\fR |
[\fB\-C\fR \fIFILENAME:FILESIZE\fR]
[\fB\-E\fR \fIFILENAME\fR]
[\fB\-F\fR \fIIMAGE_FOLDER\fR]
[\fB\-R\fR \fICURRENT_NAME:NEW_NAME\fR]
[\fB\-c\fR \fICMD\fR]
[\fB\-f\fR \fIFILE|PARAM\fR]
[\fB\-g\fR \fILOG_LEVEL\fR]
[\fB\-h\fR \fIHOST\fR]
[\fB\-i\fR \fIID[:LUN]\fR
[\fB\-n\fR \fINAME\fR]
[\fB\-p\fR \fIPORT\fR]
[\fB\-r\fR \fIRESERVED_IDS\fR]
[\fB\-t\fR \fITYPE\fR]
[\fB\-x\fR \fICURRENT_NAME:NEW_NAME\fR]
[\fB\-z\fR \fILOCALE\fR]
.SH DESCRIPTION
.B rasctl
sends commands to the rascsi process to make configuration adjustments at runtime or to check the status of the devices.

Either the -i or -l option should be specified at one time. Not both. 

You do NOT need root privileges to use rasctl.

Note: The command and type arguments are case insensitive. Only the first letter of the command/type is evaluated by the tool.

.SH OPTIONS
.TP
.BR \-C\fI " "\fIFILENAME:FILESIZE
Create an image file in the default image folder with the specified name and size in bytes.
.TP
.BR \-D\fI
Detach all devices.
.TP
.BR \-E\fI " " \fIFILENAME
Display information on an image file.
.TP
.BR \-F\fI " "\fIIMAGE_FOLDER
Set the default image folder.
.TP
.BR \-I\fI
Gets the list of reserved device IDs.
.TP
.BR \-L\fI " "\fILOG_LEVEL
Set the rascsi log level (trace, debug, info, warn, err, off).
.TP
.BR \-h\fI " " \fIHOST
The rascsi host to connect to, default is 'localhost'.
.TP
.BR \-e\fI
List all images files in the default image folder.
.TP
.BR \-N\fI
Lists all available network interfaces provided that they are up.
.TP
.BR \-O\fI
Display the available rascsi server log levels and the current log level.
.TP
.BR \-P\fI
Prompt for the access token in case rascsi requires authentication.
.TP
.BR \-l\fI
List all of the devices that are currently being emulated by RaSCSI, as well as their current status.
.TP
.BR \-m\fI
List all file extensions recognized by RaSCSI and the device types they map to.
.TP
.BR \-o\fI
Display operation meta data information.
.TP
.BR \-R\fI " "\fICURRENT_NAME:NEW_NAME
Rename an image file in the default image folder.
.TP
.BR \-p\fI " " \fIPORT
The rascsi port to connect to, default is 6868.
.TP
.BR \-r\fI " " \fIRESERVED_IDS
Comma-separated list of IDs to reserve. Pass an empty list in order to not reserve anything.
.TP
.BR \-s\fI
Display server-side settings like available images or supported device types.
.TP
.BR \-T\fI
Display all device types and their properties.
.TP
.BR \-v\fI " " \fI
Display the rascsi server version.
.TP
.BR \-V\fI " " \fI
Display the rasctl version.
.TP
.BR \-X\fI " " \fI
Shut down the rascsi process.
.TP
.BR \-d\fI " "\fIFILENAME
Delete an image file in the default image folder.
.TP
.BR \-x\fI " "\fICURRENT_NAME:NEW_NAME
Copy an image file in the default image folder.
.TP
.BR \-z\fI " "\fILOCALE
Overrides the default locale for client-facing error messages.
.TP
.BR \-i\fI " " \fIID[:LUN]
The SCSI ID and optional LUN that you want to control. (0-7:0-31)
.TP 
.BR \-c\fI " " \fICMD
Command is the operation being requested. Options are:
   a(ttach): Attach disk
   d(etach): Detach disk
   i(nsert): Insert media (removable media devices only)
   e(ject): Eject media (removable media devices only)
   p(rotect): Write protect the medium (not for CD-ROMs, which are always read-only)
   u(nprotect): Remove write protection from the medium (not for CD-ROMs, which are always read-only)
   s(how): Display device information
.IP
eject, protect and unprotect are idempotent.
.TP 
.BR \-b\fI " " \fIBLOCK_SIZE
The optional block size, either 512, 1024, 2048 or 4096 bytes. The default size is 512 bytes.
.TP
.BR \-f\fI " " \fIFILE|PARAM
Device-specific: Either a path to a disk image file, or a parameter for a non-disk device. See the rascsi(1) man page for permitted file types.
.TP 
.BR \-t\fI " " \fITYPE
Specifies the device type. This type overrides the type derived from the file extension of the specified image. See the rascsi(1) man page for the available device types. For some types there are shortcuts (only the first letter is required):
   hd: SCSI hard disk drive
   rm: SCSI removable media drive
   cd: CD-ROM
   mo: Magneto-Optical disk
   lp: SCSI printer
   bridge: Bridge device (Only applicable to the Sharp X68000)
   daynaport: DaynaPort network adapter
   services: Host services device
.TP 
.BR \-n\fI " " \fIVENDOR:PRODUCT:REVISION
The vendor, product and revision for the device, to be returned with the INQUIRY data. A complete set of name components must be provided. VENDOR may have up to 8, PRODUCT up to 16, REVISION up to 4 characters. Padding with blanks to the maxium length is automatically applied. Once set the name of a device cannot be changed.
.TP 
.BR \-u\fI " " \fIUNIT
Unit number (0-31). This will default to 0. This option is only used when there are multiple SCSI devices on a shared SCSI controller. (This is not common)

.SH EXAMPLES
Show a listing of all of the SCSI devices and their current status.
   rasctl -l


Example output:
   +----+-----+------+-------------------------------------
   | ID | LUN | TYPE | IMAGE FILE
   +----+-----+------+-------------------------------------
   |  0 |   1 | SCHD | /home/pi/harddisk.hda
   +----+-----+------+-------------------------------------

Request the RaSCSI process to attach a disk (assumed) to SCSI ID 0 with the contents of the file system image "HDIIMAGE0.HDS".
   rasctl -i 0 -f HDIIMAGE0.HDS

.SH SEE ALSO
rascsi(1), scsimon(1), rasdump(1)

Full documentation is available at: <https://www.github.com/akuker/RASCSI/wiki/>


src/x68k/RASCTL/RASCTL.C
//---------------------------------------------------------------------------
//
//	SCSI Target Emulator RaSCSI (*^..^*)
//	for Raspberry Pi
//	Powered by XM6 TypeG Technology.
//
//	Copyright (C) 2016-2021 GIMONS(Twitter:@kugimoto0715)
//
//	[ Sending commands ]
//
//---------------------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/dos.h>
#include <sys/scsi.h>
#include <iocslib.h>

//---------------------------------------------------------------------------
//
//	Primitive definitions
//
//---------------------------------------------------------------------------
typedef unsigned int UINT;
typedef unsigned char BYTE;
typedef unsigned short WORD;
typedef unsigned long DWORD;
typedef unsigned long long UL64;
typedef int BOOL;
typedef char TCHAR;
typedef char *LPTSTR;
typedef const char *LPCTSTR;
typedef const char *LPCSTR;

#if !defined(FALSE)
#define FALSE		0
#endif

#if !defined(TRUE)
#define TRUE		1
#endif

//---------------------------------------------------------------------------
//
//	Constant definitions
//
//---------------------------------------------------------------------------
#define ENOTCONN	107

//---------------------------------------------------------------------------
//
//	Struct definitions
//
//---------------------------------------------------------------------------
typedef struct
{
	char DeviceType;
	char RMB;
	char ANSI_Ver;
	char RDF;
	char AddLen;
	char RESV0;
	char RESV1;
	char OptFunc;
	char VendorID[8];
	char ProductID[16];
	char FirmRev[4];
} INQUIRY_T;

typedef struct
{
	INQUIRY_T info;
	char options[8];
} INQUIRYOPT_T;

//---------------------------------------------------------------------------
//
//	Global variables
//
//---------------------------------------------------------------------------
int scsiid = -1;

//---------------------------------------------------------------------------
//
//	Search for bridge device
//
//---------------------------------------------------------------------------
BOOL SearchRaSCSI()
{
	int i;
	INQUIRYOPT_T inq;

	for (i = 7; i >= 0; i--) {
		// Search for bridge device
		if (S_INQUIRY(sizeof(INQUIRY_T) , i, (struct INQUIRY*)&inq) < 0) {
			continue;
		}

		if (memcmp(&(inq.info.ProductID), "RASCSI BRIDGE", 13) != 0) {
			continue;
		}

		// Check if option functions are initialized
		if (S_INQUIRY(sizeof(INQUIRYOPT_T) , i, (struct INQUIRY*)&inq) < 0) {
			continue;
		}
		
		if (inq.options[0] != '1') {
			continue;
		}

		// Confirm SCSI ID
		scsiid = i;
		return TRUE;
	}

	return FALSE;
}

//---------------------------------------------------------------------------
//
//	Get messages
//
//---------------------------------------------------------------------------
BOOL ScsiGetMessage(BYTE *buf)
{
	unsigned char cmd[10];
	unsigned char sts;
	unsigned char msg;
	
	if (S_SELECT(scsiid) != 0) {
		return FALSE;
	}

	cmd[0] = 0x28;
	cmd[1] = 0;
	cmd[2] = 0;
	cmd[3] = 0;
	cmd[4] = 0;
	cmd[5] = 0;
	cmd[6] = 0;
	cmd[7] = 0x04;
	cmd[8] = 0;
	cmd[9] = 0;
	if (S_CMDOUT(10, cmd) != 0) {
		return FALSE;
	}

	if (S_DATAIN_P(1024, buf) != 0) {
		return FALSE;
	}

	S_STSIN(&sts);
	S_MSGIN(&msg);

	return TRUE;
}

//---------------------------------------------------------------------------
//
//	Send messages
//
//---------------------------------------------------------------------------
BOOL ScsiSendMessage(BYTE *buf)
{
	unsigned char cmd[10];
	unsigned char sts;
	unsigned char msg;
	
	if (S_SELECT(scsiid) != 0) {
		return FALSE;
	}

	cmd[0] = 0x2a;
	cmd[1] = 0;
	cmd[2] = 0;
	cmd[3] = 0;
	cmd[4] = 0;
	cmd[5] = 0;
	cmd[6] = 0;
	cmd[7] = 0x04;
	cmd[8] = 0;
	cmd[9] = 0;
	if (S_CMDOUT(10, cmd) != 0) {
		return FALSE;
	}

	S_DATAOUT_P(1024, buf);

	S_STSIN(&sts);
	S_MSGIN(&msg);
	
	return TRUE;
}

//---------------------------------------------------------------------------
//
//	Send commands
//
//---------------------------------------------------------------------------
BOOL SendCommand(char *buf)
{
	BYTE message[1024];
	
	memset(message, 0x00, 1024);
	memcpy(message, buf, strlen(buf));
	
	if (ScsiSendMessage(message)) {
		memset(message, 0x00, 1024);
		if (ScsiGetMessage(message)) {
			printf("%s", (char*)message);
			return TRUE;
		}
	}

	return FALSE;
}

//---------------------------------------------------------------------------
//
//	Main process
//
//---------------------------------------------------------------------------
void main(int argc, char* argv[])
{
	int opt;
	int id;
	int un;
	int cmd;
	int type;
	char *file;
	int len;
	char *ext;
	char buf[BUFSIZ];

	id = -1;
	un = 0;
	cmd = -1;
	type = -1;
	file = NULL;

	// Display help
	if (argc < 2) {
		fprintf(stderr, "SCSI Target Emulator RaSCSI Controller\n");
		fprintf(stderr,
			"Usage: %s -i ID [-u UNIT] [-c CMD] [-t TYPE] [-f FILE]\n",
			argv[0]);
		fprintf(stderr, " where  ID := {0|1|2|3|4|5|6|7}\n");
		fprintf(stderr, "        UNIT := {0|1} default setting is 0.\n");
		fprintf(stderr, "        CMD := {attach|detach|insert|eject|protect}\n");
		fprintf(stderr, "        TYPE := {hd|mo|cd|bridge}\n");
		fprintf(stderr, "        FILE := image file path\n");
		fprintf(stderr, " CMD is 'attach' or 'insert' and FILE parameter is required.\n");
		fprintf(stderr, "\n");
		fprintf(stderr, "Usage: %s -l\n\n", argv[0]);
		fprintf(stderr, "       Print device list.\n");
		fprintf(stderr, "\n");
		fprintf(stderr, "Usage: %s --stop\n\n", argv[0]);
		fprintf(stderr, "       Stop rascsi prosess.\n");
		fprintf(stderr, "\n");
		fprintf(stderr, "Usage: %s --shutdown\n\n", argv[0]);
		fprintf(stderr, "       Shutdown raspberry pi.\n");
		exit(0);
	}

	// Look for bridge device
	if (!SearchRaSCSI()) {
		fprintf(stderr, "Error: bridge not found\n");
		exit(ENOTCONN);
	}

	// Argument parsing
	opterr = 0;
	while ((opt = getopt(argc, argv, "i:u:c:t:f:l-:")) != -1) {
		switch (opt) {
			case 'i':
				id = optarg[0] - '0';
				break;

			case 'u':
				un = optarg[0] - '0';
				break;

			case 'c':
				switch (optarg[0]) {
					case 'a':				// ATTACH
					case 'A':
						cmd = 0;
						break;
					case 'd':				// DETACH
					case 'D':
						cmd = 1;
						break;
					case 'i':				// INSERT
					case 'I':
						cmd = 2;
						break;
					case 'e':				// EJECT
					case 'E':
						cmd = 3;
						break;
					case 'p':				// PROTECT
					case 'P':
						cmd = 4;
						break;
				}
				break;

			case 't':
				switch (optarg[0]) {
					case 's':				// HD(SASI)
					case 'S':
					case 'h':				// HD(SCSI)
					case 'H':
						type = 0;
						break;
					case 'm':				// MO
					case 'M':
						type = 2;
						break;
					case 'c':				// CD
					case 'C':
						type = 3;
						break;
					case 'b':				// BRIDGE
					case 'B':
						type = 4;
						break;
				}
				break;

			case 'f':
				file = optarg;
				break;

			case 'l':
				sprintf(buf, "list\n");
				SendCommand(buf);
				exit(0);

			case '-':
				if (strcmp(optarg, "shutdown") == 0) {
					sprintf(buf, "shutdown\n");
					SendCommand(buf);
					exit(0);
				} else if (strcmp(optarg, "stop") == 0) {
					sprintf(buf, "stop\n");
					SendCommand(buf);
					exit(0);
				}
				break;
		}
	}

	// ID check
	if (id < 0 || id > 7) {
		fprintf(stderr, "Error : Invalid ID\n");
		exit(EINVAL);
	}

	// Unit check
	if (un < 0 || un > 1) {
		fprintf(stderr, "Error : Invalid UNIT\n");
		exit(EINVAL);
	}

	// Command check
	if (cmd < 0) {
		cmd = 0;	// The default is 'attach'
	}

	// Type check
	if (cmd == 0 && type < 0) {

		// Attempt type detection from extension
		len = file ? strlen(file) : 0;
		if (len > 4 && file[len - 4] == '.') {
			ext = &file[len - 3];
			if (strcasecmp(ext, "hdf") == 0 ||
				strcasecmp(ext, "hds") == 0 ||
				strcasecmp(ext, "hdn") == 0 ||
				strcasecmp(ext, "hdi") == 0 ||
				strcasecmp(ext, "nhd") == 0 ||
				strcasecmp(ext, "hda") == 0) {
				// HD(SASI/SCSI)
				type = 0;
			} else if (strcasecmp(ext, "mos") == 0) {
				// MO
				type = 2;
			} else if (strcasecmp(ext, "iso") == 0) {
				// CD
				type = 3;
			}
		}

		if (type < 0) {
			fprintf(stderr, "Error : Invalid type\n");
			exit(EINVAL);
		}
	}

	// File check (the command is 'attach' with type HD)
	if (cmd == 0 && type >= 0 && type <= 1) {
		if (!file) {
			fprintf(stderr, "Error : Invalid file path\n");
			exit(EINVAL);
		}
	}

	// File check (the command is 'insert')
	if (cmd == 2) {
		if (!file) {
			fprintf(stderr, "Error : Invalid file path\n");
			exit(EINVAL);
		}
	}

	// Useless types are set to 0
	if (type < 0) {
		type = 0;
	}

	// Create command to send
	sprintf(buf, "%d %d %d %d %s\n", id, un, cmd, type, file ? file : "-");
	if (!SendCommand(buf)) {
		exit(ENOTCONN);
	}

	// Quit
	exit(0);
}
